### PR TITLE
6pm-2 create frontend pages to add and display room slots

### LINKED ISSUE #244: Create a frontend page to display Room Slot Entries
# User Story

As a user I can view the current time slots associated with physical rooms, so that we can keep track of which rooms are being used at which times and associate office hours with these slots (see #245).

# Acceptance Criteria
- [x] The navbar allows us to navigate to the Room Slots page.
- [x] There is a table showing information on all current Room Slots.

# Implementation Todos

Front end:
- [x] Add a route to the new page
- [x] Add Room Slots to navbar
- [x] Add a page component for Room Slots
- [x] Add a table on the Room Slots page to display current room slot entries
- [x] Add all necessary API calls to connect page to backend

Back end:
- No additional backend functionality is required for this story specifically (CRUD endpoints are being implemented in a different story)

Testing:
- [x] Front end tests pass and there is adequate coverage

### LINKED ISSUE #248: Create a page to add new Room Slots to the database
# User Story
As an admin I can add time slots associated with physical rooms, so that we can keep track of which rooms are being used at which times and associate office hours with these slots (see #245).

# Acceptance Criteria
- [x] There is a page dedicated to adding Room Slot entries.
- [x] We can navigate from the main Room Slots page to the New Room Slots page.
- [x] We can add Room Slot entries using a form on the page.
- [x] Room Slot entries are stored in the database using CRUD operations (blocked by #219)

# Implementation Todos

Front end:
- [x] Add a route to the new page
- [x] Add a page component for New Room Slots
- [x] Add a button linking to the New Room Slots page (see #248)
- [x] Add a form to input data for the new room slot
- [x] Add all necessary API calls to connect page to backend

Back end:
- No additional backend functionality is required for this story specifically (CRUD endpoints are being implemented in a different story)

Testing:
- [x] Front end tests pass and there is adequate coverage

### PR DESCRIPTION
**Pull request by rowantran**
In this PR we add two pages to the frontend, enabling users to display a table with current physical room slots and add new ones.
Closes #244. Closes #248.

### CODE REVIEW

**Comment by RaghubirChimni** [comment]
Need to add push some things before we can merge this into main

File: javascript/src/main/components/RoomSlots/RoomSlotForm.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            // if(createRoomSlot)
+                createRoomSlot({...roomSlot,
+                    quarter: formattedQuarter, 
+                    dayOfWeek: formattedDayOfWeek,
+                    startTime: formattedStartTime,
+                    endTime: formattedEndTime
+                });
+            
+            // else
+            //     updateRoomSlot({...roomSlot, quarter: formatedQuarter});
```
Comment (btk5h) [inline javascript/src/main/components/RoomSlots/RoomSlotForm.js:55]: Can we remove the commented code here?

**Review by btk5h — CHANGES_REQUESTED** [review]

**Review by noahcorona44 — DISMISSED** [review]
LGTM

**Review by arunryu — APPROVED** [review]
LGTM

File: javascript/src/main/components/RoomSlots/RoomSlotForm.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            if(roomSlot.quarter.substr(0, 1)==="S"){
+                quarterNum =  "2";
+            }
+            else if(roomSlot.quarter.substr(0, 1)==="M"){
+                quarterNum =  "3";
+            }
+            else if(roomSlot.quarter.substr(0, 1)==="F"){
+                quarterNum =  "4";
+            }
+            let formattedQuarter = "20" + roomSlot.quarter.substr(1, 2) + quarterNum;
```
Comment (btk5h) [inline javascript/src/main/components/RoomSlots/RoomSlotForm.js:37]: This logic can probably be refactored a bit and moved with the rest of the quarter format utilities.

File: javascript/src/main/components/RoomSlots/RoomSlotForm.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
+import React, { useState, useRef } from "react";
+import { Form, Button, Row, Col } from "react-bootstrap";
+import {checkTime, checkFilled, checkCourseQuarter} from "main/utils/FormHelpers";
+import { formatTimeToLocalTime } from "main/utils/RoomSlotFormHelpers";
+
+const RoomSlotForm = ({ createRoomSlot/*, updateRoomSlot, existingRoomSlot*/}) =>{
```
Comment (btk5h) [inline javascript/src/main/components/RoomSlots/RoomSlotForm.js:6]: Can we remove the comment here?

File: javascript/src/main/components/RoomSlots/RoomSlotTable.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import { toAMPMFormat } from "main/utils/RoomSlotTableHelpers";
+
+/**
+ * Displays the list of room slots. Code adapted from TutorTable.
+ */
+const RoomSlotTable = ({ roomSlots }) => {
+    /*
+    const AdminRenderWrapper = ({children }) => {
+        return (admin && children);
+    };*/
```
Comment (btk5h) [inline javascript/src/main/components/RoomSlots/RoomSlotTable.js:14]: Can we remove the commented code here?

File: javascript/src/main/pages/RoomSlots/RoomSlots.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  const history = useHistory();
+
+  const { data: roomSlotList, error, mutate: mutateRoomSlots } = useSWR(
+    ["/api/public/roomslot", getToken],
+    fetchWithToken
+  );
+
+  useEffect(() => {
+    mutateRoomSlots();
+  }, [mutateRoomSlots]);
```
Comment (btk5h) [inline javascript/src/main/pages/RoomSlots/RoomSlots.js:23]: What is the purpose of this piece of code? It's not obvious what you're trying to accomplish here as I don't think the useSWR mutation function was intended to be used like this.

File: javascript/src/main/pages/RoomSlots/RoomSlots.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  const isAdmin =
+    roleInfo && roleInfo.role && roleInfo.role.toLowerCase() === "admin";
```
Comment (btk5h) [inline javascript/src/main/pages/RoomSlots/RoomSlots.js:26]: I believe this long series of checks can be simplified to something like:
```suggestion
  const isAdmin = roleInfo?.role?.toLowerCase() === "admin";
```

File: javascript/src/main/utils/RoomSlotFormHelpers.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+// Takes a time in a format like XX:XXAM and converts it to HH:MM:SS format
+// to be turned into a LocalTime.
+const formatTimeToLocalTime = (time) => {
```
Comment (btk5h) [inline javascript/src/main/utils/RoomSlotFormHelpers.js:3]: I don't think this needs to be addressed in this PR but it should be noted that this logic should be replaced with a proper date parsing/formatting library in a future refactoring story.

File: javascript/src/main/utils/RoomSlotTableHelpers.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+// Adapted from StackOverflow:
+// https://stackoverflow.com/questions/13898423/javascript-convert-24-hour-time-of-day-string-to-12-hour-time-with-am-pm-and-no
+const toAMPMFormat = (time) => {
```
Comment (btk5h) [inline javascript/src/main/utils/RoomSlotTableHelpers.js:3]: Similar to formatTimeToLocalTime, this doesn't have to be refactored in this PR but we should consider a dedicated library for this in the long run.

File: javascript/src/test/pages/RoomSlots/RoomSlots.test.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+import { useHistory } from "react-router-dom";
+//import { buildDeleteRoomSlot } from "main/services/RoomSlots/RoomSlotsService";
+
+jest.mock("swr");
+jest.mock("@auth0/auth0-react");
+jest.mock("main/utils/fetch");
+
+/**jest.mock("main/services/RoomSlots/RoomSlotsService", () => ({
+  buildDeleteRoomSlot: jest.fn(),
+}) ); */
```
Comment (btk5h) [inline javascript/src/test/pages/RoomSlots/RoomSlots.test.js:16]: Can we remove the commented code here?

File: javascript/src/test/pages/RoomSlots/RoomSlots.test.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  //     useHistory.mockReturnValue({
+  //       push: pushSpy
+  //     });
+  
+  //     const { getByText } = render(<OfficeHours />);
+  //     const newOfficeHourButton = getByText("New Office Hour");
+  //     userEvent.click(newOfficeHourButton);
+  
+  //     await waitFor(() => expect(pushSpy).toHaveBeenCalledTimes(1));
+  //   });
```
Comment (btk5h) [inline javascript/src/test/pages/RoomSlots/RoomSlots.test.js:139]: If these tests are outdated, can you remove them? If they're commented out due to some other issues, can you use `test.skip` instead?

**Review by btk5h — DISMISSED** [review]

**Review by noahcorona44 — DISMISSED** [review]
LGTM

**Review by mara-downing — DISMISSED** [review]
When running on heroku qa, I get a toast notification that says "Error saving room slot" when I try to submit a new room. I am using the parameters "Library, "F20", "11:00AM", "2:00PM", Tuesday.

**Review by mara-downing — APPROVED** [review]
LGTM

**Review by noahcorona44 — APPROVED** [review]
LGTM - Reapproving changes. Nice job guys!

**Pull request merged**